compiler/src/org.graalvm.compiler.loop.phases/src/org/graalvm/compiler/loop/phases/LoopPredicationPhase.java

```diff
@@ -0,0 +1,224 @@
+        if (!(condition instanceof IntegerBelowNode) || guard.isNegated()) {
+            return;
+        }
+
+        final IntegerBelowNode rangeCheck = (IntegerBelowNode) condition;
+        final ValueNode range = rangeCheck.getY();
+        if (!loop.isOutsideLoop(range) || ((IntegerStamp) range.stamp(NodeView.DEFAULT)).lowerBound() < 0) {
+            return;
+        }
+
+        final EconomicMap<Node, InductionVariable> inductionVariables = loop.getInductionVariables();
+
+        final ValueNode x = rangeCheck.getX();
+        if (!inductionVariables.containsKey(x)) {
+            return;
+        }
+
+        final StructuredGraph graph = guard.graph();
+
+        final InductionVariable iv = inductionVariables.get(x);
+
+        Long scale = null;
+
+        final InductionVariable counter = loop.counted().getCounter();
+        if (iv.isConstantScale(counter)) {
+            scale = iv.constantScale(counter);
+        }
+
+        ValueNode offset = null;
+
+        if (iv.offsetIsZero(counter)) {
+            offset = graph.unique(ConstantNode.forInt(0));
+        } else {
+            offset = iv.offsetNode(counter);
+        }
+
+        if (offset == null || scale == null || !loop.isOutsideLoop(offset)) {
+            return;
+        }
+
+        long scaleCon = scale;
+
+        replaceGuardNode(loop, guard, range, graph, scaleCon, offset);
+    }
+
+    private static void replaceGuardNode(LoopEx loop, GuardNode guard, ValueNode range, StructuredGraph graph, long scaleCon, ValueNode offset) {
+        final InductionVariable counter = loop.counted().getCounter();
+        ValueNode rangeLong = IntegerConvertNode.convert(range, StampFactory.forInteger(64), graph, NodeView.DEFAULT);
+
+        ValueNode extremumNode = counter.extremumNode(false, StampFactory.forInteger(64));
+        final GuardingNode overFlowGuard = loop.counted().createOverFlowGuard();
+        assert overFlowGuard != null || loop.counted().counterNeverOverflows();
+        if (overFlowGuard != null) {
+            extremumNode = graph.unique(new GuardedValueNode(extremumNode, overFlowGuard));
+        }
+        final ValueNode upperNode = MathUtil.add(graph, MathUtil.mul(graph, extremumNode, ConstantNode.forLong(scaleCon, graph)),
+                        IntegerConvertNode.convert(offset, StampFactory.forInteger(64), graph, NodeView.DEFAULT));
+        final LogicNode upperCond = IntegerBelowNode.create(upperNode, rangeLong, NodeView.DEFAULT);
+
+        final ValueNode initNode = IntegerConvertNode.convert(counter.initNode(), StampFactory.forInteger(64), graph, NodeView.DEFAULT);
+        final ValueNode lowerNode = MathUtil.add(graph, MathUtil.mul(graph, initNode, ConstantNode.forLong(scaleCon, graph)),
+                        IntegerConvertNode.convert(offset, StampFactory.forInteger(64), graph, NodeView.DEFAULT));
+        final LogicNode lowerCond = IntegerBelowNode.create(lowerNode, rangeLong, NodeView.DEFAULT);
+
```

Comment:
Hi @rwestrel it's a great patch that could eliminate the range check! However, I have a question about the `lowerCond`. Is it expected to compare against "0" while not the array range like C2? The `lowerGuard` should be GE 0, isn't it?

Reply:
Thanks for looking at this. The comparison in PhaseIdealLoop::rc_predicate() is:

  if (overflow) {
    // Integer expressions may overflow, do long comparison
    range = new ConvI2LNode(range);
    register_new_node(range, ctrl);
    cmp = new CmpULNode(max_idx_expr, range);
  } else {
    cmp = new CmpUNode(max_idx_expr, range);
  }
  register_new_node(cmp, ctrl);
  BoolNode* bol = new BoolNode(cmp, BoolTest::lt);

so max_idx_expr |<| range

for a range check  in the loop that follows the pattern: ... |<| range

How does the graal patch differ?

Reply:
This is used as the `upperGuard`, right? Maybe I'm not right, so the `lowerGuard` is not used to guard "min_idx >= 0"

Reply:
PhaseIdealLoop::rc_predicate() is called twice from PhaseIdealLoop::loop_predication_impl_helper(), once for the lower bound and once for the upper bound. So that code above is used for both lower and upper guards.

Reply:
Yeah, you are right. It's also compared against the array length in C2. And maybe that because it's a kind of unsigned comparison for range check, which make it no need to compare against to zero. Thanks for your more explanation!